File: LeetCode/540-single-element-in-a-sorted-array.py
###############################################################################################
# binary search's variant
###########
# Time Complexity: O(logn)
# Space Complexity: O(1)
###############################################################################################
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        l, r = 0, len(nums)-1
        while l < r:
            mid = (l+r)>>1
            if nums[mid-1] < nums[mid] < nums[mid+1]:
                return nums[mid]
            if nums[mid] > nums[mid-1]: # for everytime the mid point to the latter one
                mid += 1
            if mid % 2: # means this is at single element's left
                l = mid + 1
            else:
                r = mid - 2
        return nums[l]

File: LeetCode/540-single-element-in-a-sorted-array.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// easy
////////////////
// Time Complexity: O(logn)
// Space Complexity: O(1)
///////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
public:
    int singleNonDuplicate(vector<int>& nums) {
        int i=0, j=nums.size()-1, mid;
        while (i < j) {
            mid = (i+j) >> 1;
            if (mid % 2 == 1) mid--; // make it to be at odd position

            if (nums[mid] == nums[mid+1]) i = mid+2;
            else j = mid;
        }
        return nums[i];
    }
};